cell_locker: maybe_rehash: ignore allocation failures

bhalevy opened the pull request:
`maybe_rehash` is complimentary and is not strictly require to succeed. If it fails, it will retry on the next call, but there's no reason to throw an exception that will fail its caller, since `maybe_rehash` is called as the final step after the caller has already succeeded with its action.

Minor enhancement for the error path, no backport required.

bhalevy commented:
@kostja can you please review / assign a reviewer from the counters area?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10582/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10582/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2767/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10788/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10582/)
#### Build Details:
- Duration: 2 hr 47 min
- Builder: spider4.cloudius-systems.com

avikivity commented on cell_locking.hh:
```diff
@@ -320,15 +320,19 @@ private:
 
             _rehash_at_size = compute_rehash_at_size(new_bucket_count);
+          } catch (...) {
+            // Ignore any error
```
It's obvious you're ignoring errors, the role of the comment is to explain why that's okay.

bhalevy replied:
clarified comment and moved to top of function in v2

kostja commented:
@bhalevy please mention the connected issue and explain the problem in the commit comment

kostja commented:
See https://github.com/scylladb/scylla-enterprise/issues/4475

kostja commented:
@bhalevy I am not sure I can comfortably review this issue without a reproducing test case. I suggest you pass it on to @piodul's team for creating the test case first.

kostja commented:
@kbr-scylla points out to me that the customer case is crashing with an assertion. how is the catch-all going to solve it?

bhalevy commented on cell_locking.hh:
```diff
@@ -320,15 +320,19 @@ private:
             _rehash_at_size = compute_rehash_at_size(new_bucket_count);
+          } catch (...) {
+            // Ignore any error
+          }
```
okay, though retrying on every request would happen today as well (though after rolling back the cell_locks changes).

avikivity replied:
That's true. But then this gives the patch more value.

bhalevy replied:
done in v2

bhalevy commented:
> @kbr-scylla points out to me that the customer case is crashing with an assertion. how is the catch-all going to solve it?

It's not. As I mentioned in the description above, it's not fixing anything. It's just a minor enhancement for the error path.

bhalevy commented:
In v2 ([91833a5](https://github.com/scylladb/scylladb/commit/91833a535b5572e1626c9faa093c93ac4f71fc4e)):
- try/catch in both maybe_rehash implementations (for cells and for partitions)
- update `_rehash_at_size` in both good and bad path to delay retry.

avikivity commented on cell_locking.hh:
```diff
@@ -229,14 +229,20 @@ private:
-
+                } catch (...) {
+                }
                 _rehash_at_size = compute_rehash_at_size(new_bucket_count);
```
Please add a comment here, explaining the logic.

I can see someone reading this and saying it's an obvious bug, since new_bucket_count doesn't match _buckets.size().

avikivity replied:
Also, add a warning, otherwise we won't know that this path happens in practice.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10625/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10625/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2806/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10829/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10625/)
#### Build Details:
- Duration: 5 hr 32 min
- Builder: i-097c9907dd1e255e4 (m5ad.8xlarge)

bhalevy commented:
In v3 ([8233303](https://github.com/scylladb/scylladb/commit/82333036f39bec6fd41210d9c17921be7217ab52)):
- restricted error handling to std::bad_alloc, other exceptions will be propagated (don't know of any, but e.g. std::invalid_argument should not be ignored)
  - dropped `noexcept` spec
- added warning in catch clauses
- added comments explaining back-off decision.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10637/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10637/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2821/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10842/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10637/)
#### Failed Tests (1/31555):
- [test_ignore_dead_nodes_for_whole_dc_replace_option](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10637/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_ignore_dead_nodes_for_whole_dc_replace_option) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ignore_dead_nodes_for_whole_dc_replace_option)
#### Build Details:
- Duration: 3 hr 0 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10850/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10850/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11008/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2985/)
#### Build Details:
- Duration: 2 hr 17 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint please consider merging
